jupyterhubutils/spawner/lsstspawner.py

```diff
@@ -0,0 +1,306 @@
+from ..utils import make_logger
+
+
+class LSSTSpawner(MultiNamespacedKubeSpawner):
+    '''This, plus the LSST Manager class structure, implements the
+    LSST-specific parts of our spawning requirements.
+    '''
+    lsst_mgr = None
+    delete_grace_period = 5
+    # In our LSST setup, there is a "provisionator" user, uid/gid 769,
+    #  that is who we should start as.
+    uid = 769
+    gid = 769
+    # The fields need to be defined; we don't use them.
+    fs_gid = None
+    supplemental_gids = []
+    extra_labels = {}
+    extra_annotations = []
+    image_pull_secrets = None
+    privileged = False
+    working_dir = None
+    lifecycle_hooks = {}  # This one will be useful someday.
+    init_containers = []
+    lab_service_account = None
+    extra_container_config = None
+    extra_pod_config = None
+    extra_containers = []
+
+    delete_namespace_on_stop = Bool(
+        True,
+        config=True,
+        help='''
+        If True, the entire namespace will be deleted when the lab pod stops.
+        Set delete_namespaced_pvs_on_stop to True if you also want to
+        delete shadow PVs created.
+        '''
+    ).tag(config=True)
+
+    enable_namespace_quotas = Bool(
+        True,
+        config=True,
+        help='''
+        If True, will create a ResourceQuota object by calling
+        `self.quota_mgr.get_resource_quota_spec()` and create a quota with
+        the resulting specification within the namespace.
+
+        A subclass should override the quota manager's
+        get_resource_quota_spec() to create a
+        situationally-appropriate resource quota spec.
+        '''
+    ).tag(config=True)
+
+    def __init__(self, *args, **kwargs):
+        self.log = make_logger()
+        self.log.debug("Creating LSSTSpawner.")
+        super().__init__(*args, **kwargs)
+    # Our API and our RBAC API are set in the super() __init__()
+    # We assume that we're using an LSST Authenticator, which will
+    #  therefore have an LSST MiddleManager.
+    #
+    # This might change with Argo Workflow.
+
+    @property
+    def options_form(self):
+        '''Present an LSST-tailored options form.'''
+        # Weird place to stitch up the LSST Manager, huh?
+        # This is the last place we can do it, because we need one for
+        #  the options form.
+        self.log.info("Updating LSSR manager with spawner information.")
+        self._set_lsst_mgr()
+        # We may want to know about the user's resource limits for the
+        #  option form.
+        self.lsst_mgr.quota_mgr.set_custom_user_resources()
+        return self.lsst_mgr.optionsform_mgr.lsst_options_form()
+
+    def _set_lsst_mgr(self):
+        self.log.info("Setting LSST Manager from authenticated user.")
+        lm = self.user.authenticator.lsst_mgr
+        self.lsst_mgr = lm
+        lm.spawner = self
+        lm.user = self.user
+        lm.username = self.user.escaped_name
+        lm.api = self.api
+        lm.rbac_api = self.rbac_api
```

Comment:
Maybe make some methods on lsst_mgr to stuff these in because this is pretty important